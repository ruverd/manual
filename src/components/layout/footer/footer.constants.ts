import facebook from '@/assets/facebook.svg';
import google from '@/assets/google.svg';
import logo from '@/assets/manual-logo.svg';
import twitter from '@/assets/twitter.svg';

import { FooterLinkType } from './footer.types';

export const COMPONENT_MAP = {
  pageLinks: {
    [FooterLinkType.PRODUCT]: {
      title: 'Product',
      links: [
        {
          text: 'Popular',
          href: '/internal/popular'
        },
        {
          text: 'Trending',
          href: '/internal/trending'
        },
        {
          text: 'Guided',
          href: '/internal/guided'
        },
        {
          text: 'Products',
          href: '/internal/products'
        }
      ]
    },
    [FooterLinkType.COMPANY]: {
      title: 'Company',
      links: [
        {
          text: 'Press',
          href: '/internal/press'
        },
        {
          text: 'Mission',
          href: '/internal/mission'
        },
        {
          text: 'Strategy',
          href: '/internal/strategy'
        },
        {
          text: 'About',
          href: '/internal/about'
        }
      ]
    },
    [FooterLinkType.INFO]: {
      title: 'Info',
      links: [
        {
          text: 'Support',
          href: '/internal/support'
        },
        {
          text: 'Customer Service',
          href: '/internal/customer-service'
        },
        {
          text: 'Get Started',
          href: '/internal/get-started'
        }
      ]
    },
    [FooterLinkType.SOCIAL]: {
      title: 'Follow Us',
      links: [
        {
          image: {
            src: facebook,
            alt: 'Facebook'
          },
          href: 'https://www.facebook.com/MenofManual'
        },
        {
          image: {
            src: google,
            alt: 'Google'
          },
          href: 'https://www.google.com'
        },
        {
          image: {
            src: twitter,
            alt: 'Twitter'
          },
          href: 'https://x.com/MenofManual'
        }
      ]
    }
  },
  logo: {
    src: logo,
    alt: 'Manual Logo'
  },
  copyright: 'Â© 2021 Manual. All rights reserverd'
};
