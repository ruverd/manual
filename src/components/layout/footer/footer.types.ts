export const enum FooterLinkType {
  PRODUCT = 'product',
  COMPANY = 'company',
  INFO = 'info',
  SOCIAL = 'social'
}
