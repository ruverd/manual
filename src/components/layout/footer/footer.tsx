import Image from 'next/image';
import Link from 'next/link';

import { COMPONENT_MAP } from './footer.constants';
import { FooterLinkType } from './footer.types';

export const Footer = () => {
  return (
    <footer className="w-full bg-brand-100 px-6 md:px-[138px]">
      <div className="flex flex-col justify-between gap-6 py-14 md:flex-row md:gap-10 md:py-[73px]">
        <div className="flex flex-col gap-10">
          <Link href="/">
            <Image
              src={COMPONENT_MAP.logo.src}
              alt={COMPONENT_MAP.logo.alt}
              className="mx-auto size-12 object-contain md:size-[75px]"
            />
          </Link>
        </div>
        <div className="grid grid-cols-2 gap-6 md:grid-cols-4">
          {Object.entries(COMPONENT_MAP.pageLinks).map(([key, value]) => (
            <div
              key={key}
              className="flex w-full flex-col items-center gap-3 md:w-44 md:items-start md:gap-5"
            >
              <h3 className="text-[10px] font-bold uppercase leading-6 tracking-[15%] text-brand">
                {value.title}
              </h3>
              {key !== FooterLinkType.SOCIAL ? (
                <ul className="flex flex-col items-center gap-3 md:items-start md:gap-5">
                  {value.links.map(link => (
                    <li key={link?.href}>
                      <Link
                        href={link.href}
                        className="text-sm leading-7 md:text-base"
                      >
                        {'text' in link ? link.text : null}
                      </Link>
                    </li>
                  ))}
                </ul>
              ) : (
                <ul className="flex flex-row items-center gap-5">
                  {value.links.map(link => (
                    <li key={link?.href}>
                      <Link href={link.href}>
                        {'image' in link && (
                          <Image
                            src={link.image.src ?? ''}
                            alt={link.image.alt ?? ''}
                            className="size-4 object-contain md:size-6"
                          />
                        )}
                      </Link>
                    </li>
                  ))}
                </ul>
              )}
            </div>
          ))}
        </div>
      </div>
      <div className="flex flex-col items-center justify-center gap-6 border-t border-brand-200 py-6 md:gap-10">
        <span className="text-sm leading-8 text-brand-800 md:text-base">
          {COMPONENT_MAP.copyright}
        </span>
      </div>
    </footer>
  );
};
