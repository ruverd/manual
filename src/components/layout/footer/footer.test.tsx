import { render, screen } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import { Footer } from '..';
import { COMPONENT_MAP } from './footer.constants';
import { FooterLinkType } from './footer.types';

vi.mock('next/image', () => ({
  default: vi.fn(({ alt, className }) => (
    <img alt={alt} className={className} data-testid="mock-image" />
  ))
}));

const renderComponent = () => {
  return render(<Footer />);
};

describe('Footer', () => {
  beforeEach(() => {
    renderComponent();
  });

  it('should render the logo image', () => {
    expect(screen.getByAltText(COMPONENT_MAP.logo.alt)).toBeInTheDocument();
  });

  it('should render all section titles', () => {
    Object.values(COMPONENT_MAP.pageLinks).forEach(section => {
      expect(screen.getByText(section.title)).toBeInTheDocument();
    });
  });

  it('should render all regular links correctly', () => {
    Object.entries(COMPONENT_MAP.pageLinks).forEach(([key, section]) => {
      if (key !== FooterLinkType.SOCIAL) {
        section.links.forEach(link => {
          if ('text' in link) {
            const linkElement = screen.getByText(link.text);
            expect(linkElement).toBeInTheDocument();
            expect(linkElement.closest('a')).toHaveAttribute('href', link.href);
          }
        });
      }
    });
  });

  it('should render all social media icons', () => {
    const socialSection = COMPONENT_MAP.pageLinks[FooterLinkType.SOCIAL];
    socialSection.links.forEach(link => {
      if ('image' in link) {
        expect(screen.getByAltText(link.image.alt ?? '')).toBeInTheDocument();
      }
    });
  });

  it('should render the copyright text', () => {
    expect(screen.getByText(COMPONENT_MAP.copyright)).toBeInTheDocument();
  });

  it('should match snapshots', () => {
    const { container } = render(<Footer />);
    expect(container).toMatchSnapshot();
  });
});
