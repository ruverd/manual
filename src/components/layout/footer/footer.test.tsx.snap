// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Footer > should match snapshots 1`] = `
<div>
  <footer
    class="w-full bg-brand-100 px-6 md:px-[138px]"
  >
    <div
      class="flex flex-col justify-between gap-6 py-14 md:flex-row md:gap-10 md:py-[73px]"
    >
      <div
        class="flex flex-col gap-10"
      >
        <a
          href="/"
        >
          <img
            alt="Manual Logo"
            class="mx-auto size-12 object-contain md:size-[75px]"
            data-testid="mock-image"
          />
        </a>
      </div>
      <div
        class="grid grid-cols-2 gap-6 md:grid-cols-4"
      >
        <div
          class="flex w-full flex-col items-center gap-3 md:w-44 md:items-start md:gap-5"
        >
          <h3
            class="text-[10px] font-bold uppercase leading-6 tracking-[15%] text-brand"
          >
            Product
          </h3>
          <ul
            class="flex flex-col items-center gap-3 md:items-start md:gap-5"
          >
            <li>
              <a
                class="text-sm leading-7 md:text-base"
                href="/internal/popular"
              >
                Popular
              </a>
            </li>
            <li>
              <a
                class="text-sm leading-7 md:text-base"
                href="/internal/trending"
              >
                Trending
              </a>
            </li>
            <li>
              <a
                class="text-sm leading-7 md:text-base"
                href="/internal/guided"
              >
                Guided
              </a>
            </li>
            <li>
              <a
                class="text-sm leading-7 md:text-base"
                href="/internal/products"
              >
                Products
              </a>
            </li>
          </ul>
        </div>
        <div
          class="flex w-full flex-col items-center gap-3 md:w-44 md:items-start md:gap-5"
        >
          <h3
            class="text-[10px] font-bold uppercase leading-6 tracking-[15%] text-brand"
          >
            Company
          </h3>
          <ul
            class="flex flex-col items-center gap-3 md:items-start md:gap-5"
          >
            <li>
              <a
                class="text-sm leading-7 md:text-base"
                href="/internal/press"
              >
                Press
              </a>
            </li>
            <li>
              <a
                class="text-sm leading-7 md:text-base"
                href="/internal/mission"
              >
                Mission
              </a>
            </li>
            <li>
              <a
                class="text-sm leading-7 md:text-base"
                href="/internal/strategy"
              >
                Strategy
              </a>
            </li>
            <li>
              <a
                class="text-sm leading-7 md:text-base"
                href="/internal/about"
              >
                About
              </a>
            </li>
          </ul>
        </div>
        <div
          class="flex w-full flex-col items-center gap-3 md:w-44 md:items-start md:gap-5"
        >
          <h3
            class="text-[10px] font-bold uppercase leading-6 tracking-[15%] text-brand"
          >
            Info
          </h3>
          <ul
            class="flex flex-col items-center gap-3 md:items-start md:gap-5"
          >
            <li>
              <a
                class="text-sm leading-7 md:text-base"
                href="/internal/support"
              >
                Support
              </a>
            </li>
            <li>
              <a
                class="text-sm leading-7 md:text-base"
                href="/internal/customer-service"
              >
                Customer Service
              </a>
            </li>
            <li>
              <a
                class="text-sm leading-7 md:text-base"
                href="/internal/get-started"
              >
                Get Started
              </a>
            </li>
          </ul>
        </div>
        <div
          class="flex w-full flex-col items-center gap-3 md:w-44 md:items-start md:gap-5"
        >
          <h3
            class="text-[10px] font-bold uppercase leading-6 tracking-[15%] text-brand"
          >
            Follow Us
          </h3>
          <ul
            class="flex flex-row items-center gap-5"
          >
            <li>
              <a
                href="https://www.facebook.com/MenofManual"
              >
                <img
                  alt="Facebook"
                  class="size-4 object-contain md:size-6"
                  data-testid="mock-image"
                />
              </a>
            </li>
            <li>
              <a
                href="https://www.google.com"
              >
                <img
                  alt="Google"
                  class="size-4 object-contain md:size-6"
                  data-testid="mock-image"
                />
              </a>
            </li>
            <li>
              <a
                href="https://x.com/MenofManual"
              >
                <img
                  alt="Twitter"
                  class="size-4 object-contain md:size-6"
                  data-testid="mock-image"
                />
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="flex flex-col items-center justify-center gap-6 border-t border-brand-200 py-6 md:gap-10"
    >
      <span
        class="text-sm leading-8 text-brand-800 md:text-base"
      >
        Â© 2021 Manual. All rights reserverd
      </span>
    </div>
  </footer>
</div>
`;
